mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala

```diff
@@ -0,0 +1,757 @@
+    set(miniBatchFraction, value)
+  }
+  setDefault(miniBatchFraction -> 1.0)
+
+  /**
+   * Set the standard deviation of initial coefficients.
+   * Default is 0.01.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setInitStd(value: Double): this.type = set(initStd, value)
+  setDefault(initStd -> 0.01)
+
+  /**
+   * Set the maximum number of iterations.
+   * Default is 100.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setMaxIter(value: Int): this.type = set(maxIter, value)
+  setDefault(maxIter -> 100)
+
+  /**
+   * Set the initial step size for the first step (like learning rate).
+   * Default is 1.0.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setStepSize(value: Double): this.type = set(stepSize, value)
+  setDefault(stepSize -> 1.0)
+
+  /**
+   * Set the convergence tolerance of iterations.
+   * Default is 1E-6.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setTol(value: Double): this.type = set(tol, value)
+  setDefault(tol -> 1E-6)
+
+  /**
+   * Set the solver algorithm used for optimization.
+   * Default is adamW.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setSolver(value: String): this.type = set(solver, value)
+  setDefault(solver -> AdamW)
+
+  /**
+   * Sets the value of param [[loss]].
+   * - "logisticLoss" is used to classification, label must be in {0, 1}.
+   * - "squaredError" is used to regression.
+   * Default is logisticLoss.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setLoss(value: String): this.type = set(loss, value)
+  setDefault(loss -> LogisticLoss)
+
+  override protected[spark] def train(dataset: Dataset[_]): FactorizationMachinesModel = {
+    val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE
+    train(dataset, handlePersistence)
+  }
+
+  protected[spark] def train(
+      dataset: Dataset[_],
+      handlePersistence: Boolean): FactorizationMachinesModel = instrumented { instr =>
+    val instances: RDD[OldLabeledPoint] =
+      dataset.select(col($(labelCol)), col($(featuresCol))).rdd.map {
+        case Row(label: Double, features: Vector) =>
+          OldLabeledPoint(label, features)
+      }
+
+    if (handlePersistence) instances.persist(StorageLevel.MEMORY_AND_DISK)
+
+    instr.logPipelineStage(this)
+    instr.logDataset(dataset)
+    instr.logParams(this, numFactors, fitBias, fitLinear, regParam,
+      miniBatchFraction, initStd, maxIter, stepSize, tol, solver, loss)
+
+    val numFeatures = instances.first().features.size
+    instr.logNumFeatures(numFeatures)
+
+    // initialize coefficients
+    val coefficientsSize = $(numFactors) * numFeatures +
+      (if ($(fitLinear)) numFeatures else 0) +
+      (if ($(fitBias)) 1 else 0)
+    val initialCoefficients =
+      Vectors.dense(Array.fill($(numFactors) * numFeatures)(Random.nextGaussian() * $(initStd)) ++
+        (if ($(fitLinear)) Array.fill(numFeatures)(0.0) else Array.empty[Double]) ++
```

Comment:
Nit: where you call `Array.fill(...)(0.0)` just `new Array[Double](...)`. It doesn't matter much but no need to set the array to 0 as this is already its value. I dont' feel strongly about it.

Reply:
another nit, `Array.emptyDoubleArray`

Reply:
> another nit, `Array.emptyDoubleArray`

The difference between `Array.emptyDoubleArray` and `Array.empty[Double]`?

Reply:
> another nit, `Array.emptyDoubleArray`

`Array.emptyDoubleArray` is a val, `Array.empty[Double]` will create a new object. So `Array.emptyDoubleArray` is more efficient, right?